### DISCUSSION TITLE
Compilation with GPU accelerated nodes on the HPC Niflheim

### DISCUSSION THREAD

**Discussion by JohanKoelsenDeWit** [SmileiPIC/Smilei#750] [Q&A]
This discussion is a continuation of a Q&A mistakenly opened under 'issues'. I will post my question here underneath:

Dear SMILEI team,

I'm having trouble compiling SMILEI with GPU accelerated nodes on the HPC, Niflheim’s, NVIDIA A100 nodes.
I'm not experienced with GPU accelerated nodes, and it is very likely me that is making the mistakes. However, if it is not too inconvenient, I was hoping you could help me compile SMILEI with GPU acceleration on Niflheim successfully?

I successfully compile with intel/2023a for CPU but encounter an issue when using CUDA for the A100 GPUs. After exporting the GPU compiler to nvcc and running make config=gpu_nvidia, I get the following error:
src/Params/Params.h:421:5: error: body of ‘constexpr’ function ‘static constexpr int Params::getGPUClusterWidth(int)’ not a return-statement 421 |     } |     ^ src/Params/Params.h: In static member function ‘static constexpr int Params::getGPUClusterGhostCellBorderWidth(int)’: src/Params/Params.h:469:5: error: body of ‘constexpr’ function ‘static constexpr int Params::getGPUClusterGhostCellBorderWidth(int)’ not a return-statement 469 |     } |     ^ src/Params/Params.h: In static member function ‘static constexpr int Params::getGPUInterpolationClusterCellVolume(int, int)’: src/Params/Params.h:494:5: error: body of ‘constexpr’ function ‘static constexpr int Params::getGPUInterpolationClusterCellVolume(int, int)’ not a return-statement 494 |     } |     ^

I see in your installation guide that several flags must be supplied in $CXXFLAGS and $GPU_COMPILER_FLAGS, and that your environment variables for jean_zay_gpu_A100 are set to:

export CXXFLAGS="-O3 -std=c++14 -fopenmp -D__GCC_ATOMIC_TEST_AND_SET_TRUEVAL=1"
export GPU_COMPILER_FLAGS="-O3 --std=c++14 -arch=sm_80 --expt-relaxed-constexpr"
export LDFLAGS="-lcudart -lcurand -lgomp"

Neither I or the admins for Niflheim are however sure what these flags must be set to in my case.

Could you please advice me to the correct CXXFLAGS and GPU_COMPILER_FLAGS for Niflheim’s A100 partition?. :-)

Thank you for your help, and kind regards, Johan

With the kind answer:

To your question:
Could you specify what environment you are using / what modules have you loaded.
You mention compiling for CPU with intel oneapi.
For GPU you should only use the compilers provided in an nvhpc package ( currently i recomment 24.5)
Our dependencies are mostly hdf5 and openmpi, these should be compiled with nvc++ after installing nvhpc
Hdf5 is simple enough and you could install it locally following our guide, for openmpi this should be handled by your support.

Regarding flags, there is lot to be said, you should inspire yourself from the jeanzay A100 example in scripts/compile_tools/machine/ but also https://smileipic.github.io/Smilei/Use/install_linux_GPU.html
FOr starters:
-fopenmp should NOT be here , neither -lgomp
you are missing the gpu specific flags for CXXFLAGS such as -acc=gpu -gpu=cc80

so it should look like:

export LDFLAGS="-acc=gpu -gpu=cc80 -cudalib=curand "
export CXXFLAGS="-acc=gpu -gpu=cc80,fastmath -std=c++14 -lcurand -Minfo=accel -w -D__GCC_ATOMIC_TEST_AND_SET_TRUEVAL=1 -I$NVDIR/Linux_x86_64/23.11/math_libs/include/"
export GPU_COMPILER_FLAGS="-O3 --std c++14 -arch=sm_80 --expt-relaxed-constexpr -I$NVDIR/hdfsrc/install/include/"

while specifying NVDIR (or you can remove those if you have a module that exports the include and lib folders)

**Comment by JohanKoelsenDeWit** [comment]
Thank you kindly.
Here is the list of loaded modules:

1. intel-compilers/2023.1.0 
2. numactl/2.0.16-GCCcore-12.3.0
3. UCX/1.14.1-GCCcore-12.3.0
4. impi/2021.9.0-intel-compilers-2023.1.0
5. iimpi/2023a
6. Szip/2.1.1-GCCcore-12.3.0
7. HDF5/1.14.0-iimpi-2023a 
8. GCCcore/11.3.0
9. zlib/1.2.12-GCCcore-11.3.0
10. binutils/2.38-GCCcore-11.3.0
11. bzip2/1.0.8-GCCcore-11.3.0
12. ncurses/6.3-GCCcore-11.3.0
13. libreadline/8.1.2-GCCcore-11.3.0
14. Tcl/8.6.12-GCCcore-11.3.0
15. SQLite/3.38.3-GCCcore-11.3.0
16. XZ/5.2.5-GCCcore-11.3.0
17. GMP/6.2.1-GCCcore-11.3.0
18. libffi/3.4.2-GCCcore-11.3.0
19.  OpenSSL/1.1
20. Python/3.10.4-GCCcore-11.3.0
21. CUDA/12.1.1

And yes, I believe I need to install nvhpc. I will try to set the flags correctly as well.

**Reply by charlesprouveur** [reply]
So, in theory you can use nvcc (cuda) to compile CUDA code and use intel compilers for the c++ code that is run on a host (CPU).
The problem with smilei is we use openacc directives AND CUDA (for the projector) . This means we need nvc++ contained in the nvhpc package to compile our c++ code, which is not possible with intel compiler. 

You should ask support to do in order:
- install nvhpc with your gcc module (11.3 is fine) (edit: in recent versions i recommend 24.5, do NOT install 24.7, i have not tested 24.9 yet) 
- compile openmpi with nvc++
- compile hdf5 with nvc++

with that you will use your python module and we can start worrying about compilation options :)

**Reply by JohanKoelsenDeWit** [reply]
The modules are built through the EasyBuild software on Niflheim. It has, however, proven quite difficult to compile openmpi with nvc++, as many other prerequisites have to be built as well.
Do you by any chance have experience with EasyBuild? :)
Thank you for your patience as well!

**Reply by charlesprouveur** [reply]
I am afraid, i don't have any experience with EasyBuild.
What we can do as a first step, is try with the openmpi builtin nvhpc (although i expect issues at runtime) to complete the compilation process of smilei .
Once we have completed that  and we see runtime issues, we can go into the details of a local openmpi install.
Were you able to create a module of hdf5 compiled with nvc++ ?
If not you can look at https://smileipic.github.io/Smilei/Use/install_linux_GPU.html for a local install of hdf5 with nvc++

**Comment by JohanKoelsenDeWit** [comment]
Hi again,
Sorry for this late reply. I have been trying to locally compile OpenMP and HDF5 on Niflheim, but it has proven too difficult due to the setup on the HPC. However, we managed to locally compile SMILEI with GPU support on a Linux computer.

The compilation was successful, but the simulation seems unable to complete a timestep when running with GPU acceleration.

PACE_2D_1.0.txt is the namelist in txt format where the   gpu_computing = True, is set for the GPU run. We also ran the simulation with CPUs only.
Attached is the output from the CPU and the GPU run. They were run as:

CPU version is run with
mpirun -n 4 build_cpu/smilei PACE_2D_1.0_cpu.py &> out_cpu.txt

GPU version is run with
mpirun -n 1 build_nvidia/smilei PACE_2D_1.0.py &> out_gpu.txt

We set the number of patches from [16,16] to [1, 1] when running with GPU acceleration as well.

It can be seen that the simulation completes a few timesteps on the CPU run but doesn't complete any with the GPU acceleration. 7 GB is allocated to the GPU, so it seems to be initiating just fine.

Finally, nvidia_envis the one from the guide used, but we had to replace one $NVDIR with gcc to recover from a bug.

I was wondering if you by chance could see what our mistake is? 🙂

Many thanks again for your patience, and best wishes, Johan
________________________________

**Comment by charlesprouveur** [comment]
Hi Johan,

Sorry to hear about the compilation issues on your Niflheim

As for your execution, i see a couple issues. You are using currently not supported features:
    momentum_initialization = "maxwell-juettner",
    ...
    boundary_conditions = [["thermalize", "thermalize"], ["thermalize", "thermalize"]],
    
I have not tested maxwell-juettner so far so this could be an issue (or not, i will have a look if necessary  EDIT: as long as you don't use injection or moving window it should not be a problem)
The BC thermalize is currently not supported on GPU (simply a question of time, i can add it if you only need that)

To check your install you can run this small test case that runs on my laptop GPU:
[input.txt](https://github.com/user-attachments/files/17488844/input.txt)   (just in case: you should rename this input.txt into input.py, this is because github does not like uploading python script)

Best,

Charles

PS: can you be more specific on this: "Finally, nvidia_envis the one from the guide used, but we had to replace one $NVDIR with gcc to recover from a bug."

**Comment by JohanKoelsenDeWit** [comment]
Hi Charles,

Thank you kindly for your reply.
Apologies for the vagueness in the encountered bug with the compilation - I'm away from the local Linux computer, so it is a colleague that is helping me compiling SMILEI and running the comparison tests. I have asked him for a clarification on the bug.

I only need the thermalized BC. If you could add that, it would be of great help!
Do you have any recommendations for other momentum-initializations apart from maxwell-juettner that are GPU supported?

Best wishes, Johan



________________________________

**Reply by charlesprouveur** [reply]
After talking with colleagues, you should be able to use maxwell-juettner, otherwise "cold" is another possible value.
I'll try to add this quickly. I'll update you once it's done and pushed to the github repo.

Charles

**Reply by JohanKoelsenDeWit** [reply]
Hi Charles,

Thank you very much.
The error encountered with the setup of nvidia_env.sh is the following:

The way NVDIR is setup and from the documentation of nvcc:

_Specify the directory in which the default host compiler executable resides.
The host compiler executable name can be also specified to ensure that the correct host compiler is selected. In addition, driver prefix options (--input-drive-prefix, --dependency-drive-prefix, or --drive-prefix) may need to be specified, if nvcc is executed in a Cygwin shell or a MinGW shell on Windows_


However, in NVDIR there is no host compiler. So the compiler return the error:

Compiling src/Particles/nvidiaParticles.cu
Compiling src/Projector/Projector1D2OrderGPUKernelCUDAHIP.cu

Smilei/NVDIR/gcc: No such file or directory
nvcc fatal   : Failed to preprocess host compiler properties.
make: *** [makefile:374: build_nvidia/src/Particles/nvidiaParticles.o] Error 1
make: *** Waiting for unfinished jobs....
Smilei/NVDIR/gcc: No such file or directory
nvcc fatal   : Failed to preprocess host compiler properties.
make: *** [makefile:374: build_nvidia/src/Projector/Projector1D2OrderGPUKernelCUDAHIP.o] Error 1

It was fixed by changing
GPU_COMPILER_FLAGS=”--compiler_bindir gcc"

Please let me know if we have misunderstood anything. 🙂

My colleague was able to run the test input, and the results from the two runs are attached here.

[gpu-test.txt](https://github.com/user-attachments/files/17504726/gpu-test.txt)
[cpu-test.txt](https://github.com/user-attachments/files/17504727/cpu-test.txt)

The GPU is a NVIDIA GeForce RTX 3090
The CPU is Intel(R) Xeon(R) W-2133 CPU @ 3.60GHz

We were however unable to see the GPU acceleration. Perhaps this is tied to the changes in the compiler flags?

**Comment by charlesprouveur** [comment]
Hi Johan,

NVDIR is an environment variable that we set ourselves at the root of the nvhpc folder we setup, ie:
export NVDIR="/local/home/myaccount/tools/nvhpc_24.5 
in which there should be two folders: "Linux_x86_64" and "modulefiles"
g++ is only used for the SMILEICXX_DEPS environment variable set in the machine file
I can imagine that gcc is not visible on your machine but that is highly unusual.

In your case, since you are using an RTX 3090, the exact options in our example work (ie -arch=sm_86)

Looking at your output there is nothing unusual, we do see "smilei will run on GPU" so on that side i think you are good.

Now onto the "acceleration" aspect, there are two things here: 

- the 1D test case is tiny, you need millions if not tens of million particle for a GPU to be "worth it". Typically i use a 3D test case with > 30e6 particles to get an idea of our performance on different hardware. It does not mean using GPU is bad per se for smaller test case, just that it is inefficient
- The RTX 3090 is a consummer card, meaning it is not made for double precision computing. As you can see in the following link
https://www.techpowerup.com/gpu-specs/geforce-rtx-3090.c3622
```
Theoretical Performance:

FP32 (float)
    35.58 TFLOPS 
FP64 (double)
    556.0 GFLOPS (1:64) 
```
Your theroretical performance is slashed by a factor 70 between single and double precision.
At this point your CPU peak performance in double precision is comparable to your GPU.

I will add another point: at high performance writing outputs (the .h5 files) will be a huge bottleneck which is why their frequency should be reduced to the minimum.  In your cases you can see that diagnostics took respectively 82% and 87% of the computing time. At this point any performance comparison between the two chips is pointless, you are only seeing the performance of your SSD.

If you want to look at pure performance you can run a much bigger test case with an output at the end for instance.

Best regards,

Charles

**Reply by JohanKoelsenDeWit** [reply]
Dear Charles,
Thank you very much for your response - I apologize for the inconvenience of my novice questions regarding the GPU computation times, your points really clarifies the results.

Moving on, I will follow your advice and try to set over a larger simulation with only outputting at the end as well.

Best wishes, Johan

**Reply by charlesprouveur** [reply]
Hi Johan,

No worries, that is perfectly normal. I am happy my explanations made it clearer for you.
On my side I am working on porting the thermal boundary conditions as there are other people interested internally :)
It's a bit tricky but it should not take long.

Best regards,
Charles

**Comment by charlesprouveur** [comment]
Hi Johan,
Thermal BC have been ported on GPU. A new version of smilei has been pushed on the github repo.
Best regards,
Charles

**Comment by JohanKoelsenDeWit** [comment]
Dear Charles,

Thank you very much for making this update. I am looking forward to using it!

Best wishes, Johan
________________________________

**Comment by smallrainbow** [comment]
I heard that vpic will have a pure gpu version coming soon. Does smilei have plans to port to gpu in the future?

**Reply by mccoys** [reply]
A good part of the features in smilei are already ported to gpu. Maybe you are referring to a particular feature?
